src/NServiceBus.Core/UnitOfWork/TransactionScopes/TransactionScopeUnitOfWork.cs

```diff
@@ -0,0 +1,77 @@
+    {
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            if (context.Settings.Get<TransportTransactionMode>() == TransportTransactionMode.TransactionScope)
+            {
+                throw new Exception("A Transaction scope unit of work can't be used when the transport already uses a scope for the receive operation. Please remove the call to config.UnitOfWork().WrapHandlersInATransactionScope() or configure the transport to use a lower transaction mode");
+            }
+
+            var transactionOptions = context.Settings.Get<Settings>().TransactionOptions;
+
+            context.Container.ConfigureComponent(b => new TransactionScopeUnitOfWorkBehavior(transactionOptions), DependencyLifecycle.InstancePerCall);
+
+            context.Pipeline.Register("HandlerTransactionScopeWrapper", typeof(TransactionScopeUnitOfWorkBehavior), "Makes sure that the handlers gets wrapped in a transaction scope");
+        }
+
+        public class Settings
+        {
+            public TransactionOptions TransactionOptions { get; }
+            
+            public Settings(TimeSpan? requestedTimeout, IsolationLevel? requestedIsolationLevel)
+            {
+
+                var timeout = TransactionManager.DefaultTimeout;
+                var isolationLevel = IsolationLevel.ReadCommitted;
+                if (requestedTimeout.HasValue)
+                {
+                    var maxTimeout = GetMaxTimeout();
+
+                    if (requestedTimeout.Value > maxTimeout)
+                    {
+                        throw new ConfigurationErrorsException(
+                            "Timeout requested is longer than the maximum value for this machine. Please override using the maxTimeout setting of the system.transactions section in machine.config");
+                    }
+
+                    timeout = requestedTimeout.Value;
+                }
+              
+                if (requestedIsolationLevel.HasValue)
+                {
+                    isolationLevel = requestedIsolationLevel.Value;
+                }
+              
+                TransactionOptions = new TransactionOptions
+                {
+                    IsolationLevel = isolationLevel,
+                    Timeout = timeout
+                };
+            }
+
+            static TimeSpan GetMaxTimeout()
+            {
+                //default is always 10 minutes
+                var maxTimeout = TimeSpan.FromMinutes(10);
+
+                var systemTransactionsGroup = ConfigurationManager.OpenMachineConfiguration()
+                    .GetSectionGroup("system.transactions");
+
+                var machineSettings = systemTransactionsGroup?.Sections.Get("machineSettings") as MachineSettingsSection;
+
+                if (machineSettings != null)
+                {
+                    maxTimeout = machineSettings.MaxTimeout;
+                }
+
+                return maxTimeout;
+            }
+        }
```

Comment:
I had a little bit of a struggle with this myself. There's also analogous class in the SQL transport, so we have 3 "copies" total.

Given that MSMQ is going to be a separate package, we're looking at sharing this between 3 NuGet packages. This is clearly not ideal. On the other hand, it allows us to evolve e.g. defaults for these packages independently if we think it's needed at some point.

Currently the only way to fix it would be to expose this out of the core, which I think is suboptimal. I think we need to live with the duplication.

@andreasohlund Thoughts about leaving this duplicated?